File: web/platform/runtime.mjs
// runtime.mjs — the loop that joins a cart, a platform backend and a world
// (DESIGN.md §11 M2, §12).
//
// Two clocks, deliberately separated:
//
//   TICKS are the world's. One tick = one `world_step`, and the only things
//   that reach it are the action set the cart returned and the input snapshot
//   taken at that tick's boundary. The tick count, the action log and the
//   resulting state are a pure function of each other (I4), which is what
//   makes a save an action log and a replay exact.
//
//   FRAMES are the renderer's. A frame may repaint several times per tick or
//   skip ticks entirely; movement interpolation, easing and cue animations are
//   presentation state and live above the line. Nothing a frame does can
//   change a fact.
//
// The wall clock is therefore allowed to decide HOW MANY ticks elapse (that is
// what the accumulator does) and never WHAT a tick contains. Under lockstep
// the accumulator is replaced by the network's tick schedule and nothing else
// about a cart changes.

/**
 * @param {object} o
 * @param {object} o.platform  from createPlatform(backend)
 * @param {object} o.backend   the same backend (for present/resize)
 * @param {object} o.cart      { tick(ctx), draw(ctx), init?(ctx) }
 * @param {object} o.world     the generated typed binding's session
 * @param {number} [o.tps]     world ticks per second
 */
/** The LIVE source: what the cart proposed is what happens. */
export const live = () => ({ orders: (proposed) => proposed, done: () => false });

/** The REPLAY source: the log decides, and the cart's proposal is discarded.
 *  A log entry is a list of ground action terms — the save format — so replay
 *  needs no binding-time argument structure and cannot re-run protocol checks
 *  that were already passed when the orders were first collected. */
export const replay = (log) => {
  let i = 0;
  return {
    orders: () => (i < log.length ? { terms: log[i++] } : null),
    done: () => i >= log.length,
  };
};

/**
 * @param {object} o
 * @param {object} [o.source]    an action source: live (default) or replay(log)
 * @param {object} [o.identity]  { story, game } — the save's compatibility key
 */
export function createRuntime({ platform, backend, cart, world, tps = 20,
                               source = live(), identity = {} }) {
  const ctx = {
    ...platform.cart,
    world,
    /** ticks elapsed — the cart's only legitimate notion of time */
    tick: 0,
    /** what the last step handed back (§12's three streams). Cleared every
     *  tick, exactly like the engine's own per-tick buffers, so a cue cannot
     *  be rendered twice by a frame that happens to run late. */
    step: { emits: [], changed: [], edges: [], rejected: null },
    /** the action log — a save is (engine-hash, game-hash, THIS) */
    log: [],
  };

  // Assets are registered before the cart runs: loading is above the durability
  // line (a backend decides what an atlas IS), naming one is below it.
  for (const [name, def] of Object.entries(cart.sheets ?? {}))
    backend.defineSheet(name, def);

  cart.init?.(ctx);

  /** Advance the world exactly one tick. */
  function step() {
    platform.sampleInput();
    ctx.step = { emits: [], changed: [], edges: [], rejected: null };

    // The cart PROPOSES; the source DISPOSES. Live, that is the same thing.
    // Replaying, the log decides and the proposal is thrown away — which is
    // what makes a save loadable rather than merely recorded, and it is the
    // seat a network source takes for lockstep.
    const proposed = cart.tick(ctx);
    const orders = source.orders(proposed, ctx.tick);
    const terms = orders?.items ? orders.items.map((a) => a.term) : orders?.terms;
    if (terms && terms.length) {
      try {
        world.stepTerms(terms);
        ctx.log.push(terms);
        ctx.step.emits = world.emits();
        ctx.step.changed = world.changed();
        ctx.step.edges = world.edges();
      } catch (e) {
        // A rejected step is a contested world or a protocol violation — the
        // world did not move, so neither does the log. Surfaced to the cart
        // rather than thrown past it: a client that cannot render "that did
        // not happen" is a client that hides authoring errors.
        ctx.step.rejected = e.message;
      }
    }
    ctx.tick++;
    cart.after?.(ctx);
  }

  /** A save is (engine-hash, game-hash, action-log) — §12. Not a state dump:
   *  state written outside the log is state replay cannot reproduce, which
   *  forfeits shareable playthroughs, branching and time travel in one move. */
  function save() {
    return { format: 1, story: identity.story ?? null,
             game: identity.game ?? null, engine: identity.engine ?? null,
             log: ctx.log.map((t) => [...t]) };
  }

  /** Replay a save into THIS world, which must be freshly opened: the log is
   *  from genesis, so loading into a played world would append a second
   *  history to the first. Refuses a save from a different story rather than
   *  replaying orders whose atoms mean something else now. */
  function load(saved) {
    if (saved.game && identity.game && saved.game !== identity.game)
      throw new Error('this save is from a different version of ' +
        (saved.story ?? 'the story') + ' — its action log names atoms this one may not have');
    if (ctx.tick !== 0 || ctx.log.length)
      throw new Error('load into a fresh world: a log is a history from genesis');
    source = replay(saved.log);
    return saved.log.length;
  }

  /** Repaint once. Never mutates the world. */
  function render() {
    backend.beginFrame?.();
    cart.draw(ctx);
    backend.present?.();
  }

  return {
    ctx,
    step,
    render,
    save,
    load,
    /** Has the current source run out? True once a replay reaches its end. */
    replaying: () => !source.done(),
    /** Run n ticks with a repaint after each — the headless/test driver. */
    advance(n = 1) { for (let i = 0; i < n; i++) { step(); render(); } },
    /** The browser loop: rAF for frames, an accumulator for ticks. */
    start() {
      const dt = 1000 / tps;
      let last = null, acc = 0, stop = false;
      const frame = (now) => {
        if (stop) return;
        if (last === null) last = now;
        acc += Math.min(250, now - last);        // a long pause is not 200 ticks
        last = now;
        while (acc >= dt) { acc -= dt; step(); }
        render();
        requestAnimationFrame(frame);
      };
      requestAnimationFrame(frame);
      return () => { stop = true; };
    },
  };
}

File: src/runtime/runtime.c
#include "runtime/runtime.h"

#include <stdlib.h>
#include <string.h>

typedef struct { uint32_t act[RT_MAX_ACTIONS]; int n; } rt_orders;

struct rt {
    plat   *p;
    world  *w;
    intern *syms;
    rt_cart cart;
    char    story[256];

    uint64_t tick;
    char     rejected[192];
    bool     has_rejected;

    /* the action log — a save is (engine-hash, game-hash, THIS) */
    rt_orders *log;
    int        nlog, caplog;

    /* the replay source: the log decides and the cart's proposal is dropped */
    rt_orders *feed;
    int        nfeed, atfeed;
    bool       replaying;
};

rt *rt_open(plat *p, world *w, intern *syms, rt_cart cart, const char *story)
{
    rt *r = calloc(1, sizeof *r);
    r->p = p;
    r->w = w;
    r->syms = syms;
    r->cart = cart;
    snprintf(r->story, sizeof r->story, "%s", story ? story : "");
    for (int i = 0; i < cart.nsheets; i++)
        plat_define_sheet(p, cart.sheets[i], NULL);
    if (cart.init) cart.init(cart.ctx, r);
    return r;
}

void rt_close(rt *r)
{
    if (!r) return;
    free(r->log);
    free(r->feed);
    free(r);
}

world  *rt_world(rt *r)    { return r->w; }
plat   *rt_platform(rt *r) { return r->p; }
intern *rt_syms(rt *r)     { return r->syms; }
uint64_t rt_tick_count(const rt *r) { return r->tick; }
const char *rt_rejected(const rt *r) { return r->has_rejected ? r->rejected : NULL; }
bool rt_replaying(const rt *r) { return r->replaying && r->atfeed < r->nfeed; }

static void log_push(rt *r, const uint32_t *act, int n)
{
    if (r->nlog == r->caplog) {
        r->caplog = r->caplog ? r->caplog * 2 : 64;
        r->log = realloc(r->log, (size_t)r->caplog * sizeof *r->log);
    }
    rt_orders *o = &r->log[r->nlog++];
    o->n = n;
    for (int i = 0; i < n; i++) o->act[i] = act[i];
}

void rt_step(rt *r)
{
    plat_sample_input(r->p);
    r->has_rejected = false;

    uint32_t proposed[RT_MAX_ACTIONS];
    int n = 0;
    if (r->cart.tick) n = r->cart.tick(r->cart.ctx, r, proposed, RT_MAX_ACTIONS);
    if (n < 0) n = 0;
    if (n > RT_MAX_ACTIONS) n = RT_MAX_ACTIONS;

    /* The cart proposes; the source disposes. Live, those are the same thing;
     * replaying, the log's entry replaces the proposal entirely — which is
     * what makes a save loadable rather than merely recorded, and it is the
     * seat a network source takes for lockstep. */
    const uint32_t *act = proposed;
    if (r->replaying) {
        if (r->atfeed < r->nfeed) {
            act = r->feed[r->atfeed].act;
            n = r->feed[r->atfeed].n;
            r->atfeed++;
        } else {
            n = 0;
        }
    }

    if (n > 0) {
        char err[160];
        if (world_step(r->w, act, n, err, sizeof err) == 0) {
            log_push(r, act, n);
        } else {
            snprintf(r->rejected, sizeof r->rejected, "%s", err);
            r->has_rejected = true;
        }
    }
    r->tick++;
    if (r->cart.after) r->cart.after(r->cart.ctx, r);
}

void rt_render(rt *r)
{
    plat_begin_frame(r->p);
    if (r->cart.draw) r->cart.draw(r->cart.ctx, r);
    plat_present(r->p);
}

void rt_advance(rt *r, int nticks)
{
    for (int i = 0; i < nticks; i++) { rt_step(r); rt_render(r); }
}

int rt_save(const rt *r, FILE *out)
{
    if (!out) return -1;
    fprintf(out, "# a save, in the only form this engine has one: the action\n"
                 "# log (DESIGN.md §12). One line per tick; a blank line is a\n"
                 "# tick that changed nothing.\n\n");
    if (r->story[0]) fprintf(out, "story %s\n\n", r->story);
    for (int i = 0; i < r->nlog; i++) {
        for (int j = 0; j < r->log[i].n; j++)
            fprintf(out, "%s%s", j ? " " : "", intern_name(r->syms, r->log[i].act[j]));
        fputc('\n', out);
    }
    return r->nlog;
}

static void feed_push(rt *r, const uint32_t *act, int n)
{
    if (r->nfeed == r->atfeed && r->feed == NULL) r->atfeed = 0;
    r->feed = realloc(r->feed, (size_t)(r->nfeed + 1) * sizeof *r->feed);
    rt_orders *o = &r->feed[r->nfeed++];
    o->n = n;
    for (int i = 0; i < n; i++) o->act[i] = act[i];
}

int rt_load(rt *r, const char *text, char *err, size_t errsz)
{
    if (r->tick != 0 || r->nlog != 0) {
        if (err) snprintf(err, errsz,
                          "load into a fresh world: a log is a history from genesis");
        return -1;
    }
    char *copy = strdup(text ? text : "");
    for (char *line = copy, *next; line && *line; line = next) {
        char *nl = strchr(line, '\n');
        next = nl ? nl + 1 : NULL;
        if (nl) *nl = '\0';
        while (*line == ' ' || *line == '\t') line++;
        if (*line == '#' || *line == '\0') continue;
        if (strncmp(line, "story ", 6) == 0) {
            /* a log replayed against the wrong world is the classic desync,
             * so the save names its story and this is a refusal, not a
             * mystery. Compared on the basename, since a save travels and an
             * absolute path does not. */
            const char *want = line + 6;
            const char *wb = strrchr(want, '/');
            const char *hb = strrchr(r->story, '/');
            wb = wb ? wb + 1 : want;
            hb = hb ? hb + 1 : r->story;
            if (r->story[0] && strcmp(wb, hb) != 0) {
                if (err) snprintf(err, errsz,
                                  "this save is from '%s' — its action log names "
                                  "atoms '%s' may not have", wb, hb);
                free(copy);
                return -1;
            }
            continue;
        }
        uint32_t act[RT_MAX_ACTIONS];
        int n = 0;
        for (char *t = strtok(line, " \t"); t && n < RT_MAX_ACTIONS;
             t = strtok(NULL, " \t"))
            act[n++] = intern_id(r->syms, t);
        feed_push(r, act, n);
    }
    free(copy);
    r->replaying = true;
    r->atfeed = 0;
    return r->nfeed;
}
